### PR TITLE
docs(parser): change parser documentation

### LINKED ISSUE #1371: Refactor Parser docs code snippets

### PR DESCRIPTION
**Pull request by anafalcao**
**Issue number:** #1371

## Summary

This PR refreshes Parser docs with newer features like examples, improves wording and navigation, focused on Powertools v3.

### Changes

* [x] Clean up the initial explanation
* [x] Clean up Key Features
* [x] Move the info "this is not necessary if using Layers" to the top
* [x] Remove Pydantic 1 integration
* [x] Add a QuickStart with BaseModels and Parse events
* [x] Add sections Getting Started and Advanced
* [x] Fix the parse() function example and initial explanation
* [x] Add differences between of parse() and event_parser
* [x] Clean up the example of extending built-in models; add pre and post explanations
* [x] Clean up the Envelopes example and improve the initial explanation
* [x] Complete the Validating fields example
* [x] Clean up the root_validator example
* [x] Move the Pydantic helper functions box to Getting started section
* [x] Restructure Serialization; add a functional example
* [x] Clean up the two examples of string fields that contain JSON data
* [x] Clean up the Bringing your own envelope examples

### User experience

> Please share what the user experience looks like before and after this change

modifcations_for_new_parser

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by anafalcao** [comment]
Hey @leandrodamascena ! 
Can you please review this PR? Something is missing for this issue https://github.com/aws-powertools/powertools-lambda-python/issues/5390, and I may need some direction to fix it.
Thanks

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Hey @anafalcao, there is something weird here. Pls take a look

image

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights here.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md:207]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md:213]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Missing highlights.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @anafalcao! This is amazing to start reviewing this PR, indeed a great improvement in our documentation.

I see we have room to improve some areas and I left some comments. Can you please review and address them before we move to the next round of this review?

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md:207]: Added highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Fixed it! It was breaking due to a "---" prior to the sub header

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Thanks! Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md:207]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md:213]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Added the highlights

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Added the highlights

**Review by anafalcao — COMMENTED** [review]
Thank you so much for the reviews @leandrodamascena ! Fixed your comments

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: We are supporting only Pydantic v2 now, so, the part "each pydanic" is not necessary anymore. What about: Parser supports Pydantic v2. To use Parser, you'll need to install the necessary dependencies for Pydantic v2 beforehand.?

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: You can remove this block. Pydantic v2 doesn't work like that.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Great example, but there is a problem here. When using `@event_parser` you can't catch the exception inside the Lambda because the validation [happens before](https://github.com/aws-powertools/powertools-lambda-python/blob/7eb6026264b434f0b1017b41b2621c165c2b5bde/aws_lambda_powertools/utilities/parser/parser.py#L110) to execute the Lambda handler. So, in this case you can remove the try/except block and add a new line before the `The @event_parser(model=MyEvent) automatically` explaining that this function fails fast.

Also, pelase create a new JSON file and add as example of the payload. Something like this: https://docs.powertools.aws.dev/lambda/python/latest/utilities/idempotency/#sample-event.

In this case your payload is just id and name:

{
"id": "bla",
"name": "bla"
}

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]:
```suggestion
The `@event_parser(model=MyEvent)` automatically parses the event into the specified Pydantic model `MyEvent`.
```

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Amazing explanation! 🚀

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Great explanation.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: same here:

>Also, pelase create a new JSON file and add as example of the payload. Something like this: [docs.powertools.aws.dev/lambda/python/latest/utilities/idempotency#sample-event](https://docs.powertools.aws.dev/lambda/python/latest/utilities/idempotency/#sample-event).

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: Please create an example in this section.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Thanks for addressing to all the feedback, @anafalcao! I have some new comments in this second round! 

Lets go!! We're almost ready to merge this PR.

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Great call, updated!

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Removed!

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Thanks, created the example!

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Fixed it! Thanks

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Thanks, created the example!

File: docs/utilities/parser.md
Comment (anafalcao) [reply docs/utilities/parser.md]: Example and sample event added to built-in models

**Review by leandrodamascena — DISMISSED** [review]
Hi @anafalcao! Congratulations on getting your first PR approved as a Powertools maintainer, this is a huge achievement in such a short time. I made a few small changes before approving it and let me tell you:

1 - We try to use active voice as much as possible, so in areas where you were saying "parser allows you to use XYZ", I changed it to "you can use XYZ in the parser utility".

2 - Some examples were importing things from Powertools instead of Pydantic, as it is better to import BaseModel directly from Pydantic than from Powertools, for example. Both work, but the customer can have more control over the imports.

3 - Mypy was complaining about some types, but don't worry, mypy is a nightmare sometimes.

4 - Some highlights were wrong and I fixed them.

Congratss!! 🚀 ⭐

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]:
```suggestion
- Compatible with Pydantic v2.x
```

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md:22]:
```suggestion
Powertools supports Pydantic v2.x, so make sure to install it before using this utility.
```

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: I think this should say `aws-lambda-powertools[parser]` - if unsure check with @leandrodamascena.

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]:
```suggestion
You can also add as a dependency in your preferred tool, e.g. `requirements.txt`, `pyproject.toml`, etc.
```

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: Is "Parse" here intentional or should it be "Parser"?

If it's intentional, feel free to resolve this thread.

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: The verb in the last sentence is incorrect, should be either "we raise" or "it raises", I think.

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: Is the "Keys" plural intentional? Should it be "Key differences" instead?

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]:
```suggestion
You can use pre-built models to work events from AWS services, so you don’t need to create them yourself. We’ve already done that for you!
```

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: I think this is a leftover from the Data Classes feature, afaik we don't use JMESPath expressions here.

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md:183]: What are the double backtick quotes (``) here? Is it to create a new line? wasn't `<br />` more known/common?

File: docs/utilities/parser.md
Comment (dreamorosi) [inline docs/utilities/parser.md]: When you bring your own Envelope model, do you need to actually reimplement the `parse` method or decorator, or we meant "using"?

Also, is `parse` it actually a method (as in class method) or a function?

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Left a few comments.

Thanks a lot for the work on this PR!

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Done

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md:22]: Powertools only = Before this, we supported both v1 and v2, so I think it's okay to emphasize the "only" world.

Dependencies needed for Pydantic v2 before using Parser = In pydantic v1, things like base-settings, email-validator and others came together in the Pydantic package instalattion, but now they are separate packages.

So I prefer to keep it as it is.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Ohhh my god! My bad! Yeah

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Nice, accepted.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: You're right, `Parser` is better.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Yeah, it raises.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Great catch

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: Nice suggestion.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: 100%! Thanks for this catch.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md:183]: This is because it is like an escape function to use `NewImage` and `OldImage` within the line.

File: docs/utilities/parser.md
Comment (leandrodamascena) [reply docs/utilities/parser.md]: The only reason it makes sense for you to bring your own Envelope is because you want to parse the payload using different fields or applying some logic, so yes, it makes sense for you to implement the `parse` method when bringing yours.

**Review by leandrodamascena — APPROVED** [review]
Hi @dreamorosi! Thanks for such a detailed review, I pushed a new commit with the last few things before we merge.

**Review by dreamorosi — APPROVED** [review]
Great work @anafalcao, looking forward to see this merged.

We should probably bring these changes to the TS docs as well.

**Pull request merged**
